### PR TITLE
[BUG] Fixed behavior of DataFrameGroupBy.apply to respect _group_selection_context

### LINKED ISSUE #28549: Many groupby tests depend on a bug in DataFrameGroupBy.apply
Discovered while fixing an issue in #28541 

A large number of tests relating to `DataFrameGroupBy.apply` depend on a bug in `apply` where the grouped column is not removed from the returned DataFrame.

Here are some examples:

- [`test_apply_chunk_view`](https://github.com/pandas-dev/pandas/blob/4edf938aedf55b9e6fbfb3199f70f857e8ec7e41/pandas/tests/groupby/test_apply.py#L366)
- [`test_apply_multiindex_fail`](https://github.com/pandas-dev/pandas/blob/4edf938aedf55b9e6fbfb3199f70f857e8ec7e41/pandas/tests/groupby/test_apply.py#L366)

There are around ~20 cases of this.

---

In general, they expect the result of something like this:

```
df = pd.DataFrame({"key": [1, 1, 1, 2, 2, 2, 3, 3, 3], "value": range(9)})
df.groupby('key').apply(lambda x: x.sum())
```

To include the "key" column in the result, which should not be the case.

```
     key  value
key
1      3      3
2      6     12
3      9     21
```

The underlying issue with `apply` is a straightforward fix, all that needs to be done is change [this return](https://github.com/pandas-dev/pandas/blob/master/pandas/core/groupby/groupby.py#L729) to use the `_group_selection_context`, but the PR will have to include updates to many tests.

### PR DESCRIPTION
**Pull request by christopherzimmerman**
- [x] closes #28549
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

This issue needs to be addressed before #28541 can be merged

### CODE REVIEW

File: pandas/core/groupby/groupby.py [hunk 1]
```diff
@@ -721,7 +721,9 @@ def f(g):
             f = func
 
         # ignore SettingWithCopy here in case the user mutates
-        with option_context("mode.chained_assignment", None):
+        with option_context(
+            "mode.chained_assignment", None
+        ) as _, _group_selection_context(self) as _:
```
Comment (WillAyd) [inline pandas/core/groupby/groupby.py]: I think just don't need the aliases at all

File: pandas/tests/groupby/test_apply.py [hunk 2]
```diff
@@ -363,11 +363,19 @@ def f(group):
         tm.assert_frame_equal(result.loc[key], f(group))
 
 
-def test_apply_chunk_view():
+@pytest.mark.parametrize("as_index", [False, True])
```
Comment (WillAyd) [inline pandas/tests/groupby/test_apply.py]: Do you even need this in the test? Looks like the test itself just remove the `True` case

File: pandas/tests/groupby/test_apply.py [hunk 3]
```diff
@@ -638,24 +656,37 @@ def test_func(x):
     tm.assert_frame_equal(result, expected)
 
 
-def test_groupby_apply_none_first():
+@pytest.mark.parametrize("as_index", [True, False])
```
Comment (WillAyd) [inline pandas/tests/groupby/test_apply.py]: Can you make `as_index` a fixture in the groupby conftest? I think will have generic utility in this set of tests

**Review by WillAyd — CHANGES_REQUESTED** [review]
Stylistic comments; need to review in more detail later

Does this have any backwards compatibility impacts on the API?

**Comment by christopherzimmerman** [comment]
Yes there are impacts. The change to [this section](https://github.com/pandas-dev/pandas/blob/master/pandas/core/groupby/groupby.py#L724) guarantees that the grouper column from `apply` will never be in the output unless `as_index` is set to `False`.  Before, it was based on whether or not the reduction operation would fail on the grouper column.

File: pandas/tests/groupby/test_groupby.py [hunk 4]
```diff
@@ -944,16 +947,28 @@ def f_no_copy(x):
...
+        result1 = df.groupby("key", group_keys=True, as_index=as_index).apply(
+            lambda x: x[:].key
+        )
+        result2 = df.groupby("key", group_keys=True, as_index=as_index).apply(
+            lambda x: x.key
+        )
+        return result1, result2
+
+    if as_index:
+        with pytest.raises(AttributeError):
```
Comment (WillAyd) [inline pandas/tests/groupby/test_groupby.py]: Can you add a `match=` argument here? Should see that fairly often elsewhere in the code

Comment (WillAyd) [reply pandas/tests/groupby/test_groupby.py]: Also this might be something that (while not previously correct) users would rely on; see top message on how best to handle

**Review by WillAyd — CHANGES_REQUESTED** [review]
This is a great change. The only thing we might need to be careful of is that users *may* have also relied on this buggy behavior in the past (to their credit, we did internally!)

To guard against that can you add a section to the whatsnew for backwards incompatible API changes that spells out the changes in more detail? I think it would be particularly worthwhile to point out the test case that previously worked but now raises an `AttributeError`

File: pandas/tests/groupby/conftest.py [hunk 5]
```diff
@@ -110,3 +110,8 @@ def reduction_func(request):
     """yields the string names of all groupby reduction functions, one at a time.
     """
     return request.param
+
+
+@pytest.fixture
+def as_index():
+    return [True, False]
```
Comment (jreback) [inline pandas/tests/groupby/conftest.py]: add a doc string

**Review by jreback — CHANGES_REQUESTED** [review]
will look soon

File: pandas/tests/groupby/test_groupby.py (hunk 4 above)
Comment (christopherzimmerman) [reply pandas/tests/groupby/test_groupby.py]: I took a stab at the whatsnew.  I followed the pattern of another breaking change, but I have never done it before so I'm sure it will need tweaks.

File: doc/source/whatsnew/v1.0.0.rst [hunk 6]
```diff
@@ -174,6 +174,43 @@ Backwards incompatible API changes
 
    pd.arrays.IntervalArray.from_tuples([(0, 1), (2, 3)])
 
+.. _whatsnew_1000.api_breaking.GroupBy.apply:
+
+``GroupBy.apply`` behaves consistently with `as_index`
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+- The result of :meth:`GroupBy.apply` sometimes contained the grouper column(s),
```
Comment (WillAyd) [inline doc/source/whatsnew/v1.0.0.rst]: Can you add an example where `GroupBy.apply` is used to essentially concat groups together? I don't think many will find this example controversial, but doing something like `df.groupby("a").apply(lambda x: x.iloc[0])` may surprise some going forward if they previously expected the grouping to be included as part of the apply output

**Review by WillAyd — CHANGES_REQUESTED** [review]

**Review by WillAyd — APPROVED** [review]
lgtm @jreback

File: pandas/tests/groupby/test_groupby.py [hunk 7]
```diff
@@ -89,9 +89,12 @@ def max_value(group):
 
     applied = df.groupby("A").apply(max_value)
     result = applied.dtypes
+
+    # GH 28549
```
Comment (jreback) [inline pandas/tests/groupby/test_groupby.py]: a comment about something that is not there is not helpful if the reader doesn't know the context, which is pretty hard when just browsing code that is not part of the change itself. remove the comment.

File: pandas/tests/groupby/test_groupby.py [hunk 8]
```diff
@@ -944,16 +947,28 @@ def f_no_copy(x):
...
 
     # GH 8467
     # first show's mutation indicator
     # second does not, but should yield the same results
     df = DataFrame({"key": [1, 1, 1, 2, 2, 2, 3, 3, 3], "value": range(9)})
 
-    result1 = df.groupby("key", group_keys=True).apply(lambda x: x[:].key)
-    result2 = df.groupby("key", group_keys=True).apply(lambda x: x.key)
-    assert_series_equal(result1, result2)
+    def run_test(df, as_index):
```
Comment (jreback) [inline pandas/tests/groupby/test_groupby.py]: do not nest functions like this; best to split this into multiple tests and/or parameterize

File: pandas/tests/groupby/test_groupby.py [hunk 9]
```diff
@@ -1341,12 +1358,19 @@ def _check_groupby(df, result, keys, field, f=lambda x: x.sum()):
...
-def test_dont_clobber_name_column():
+def test_dont_clobber_name_column(as_index):
     df = DataFrame(
         {"key": ["a", "a", "a", "b", "b", "b"], "name": ["foo", "bar", "baz"] * 2}
     )
 
-    result = df.groupby("key").apply(lambda x: x)
+    result = df.groupby("key", as_index=as_index).apply(lambda x: x)
+
+    # GH 28549
```
Comment (jreback) [inline pandas/tests/groupby/test_groupby.py]: i don't find these comments / references helpful

File: pandas/tests/groupby/test_transform.py [hunk 10]
```diff
@@ -470,16 +470,23 @@ def f(group):
...
     # this is by definition a mutating operation!
     with pd.option_context("mode.chained_assignment", None):
-        for key, group in grouped:
+        for index, (key, group) in enumerate(grouped):
             res = f(group)
-            assert_frame_equal(res, result.loc[key])
+            # GH 28549
+            # if as_index need to drop column from res
+            if as_index:
+                res = res.drop("c", 1)
```
Comment (jreback) [inline pandas/tests/groupby/test_transform.py]: make this as part of the same if clause

**Review by jreback — CHANGES_REQUESTED** [review]
your change is ok, but many comments on the test changes.

File: pandas/tests/groupby/test_function.py [hunk 11]
```diff
@@ -341,10 +349,13 @@ def test_cython_api2():
     tm.assert_frame_equal(result, expected)
 
     # GH 13994
-    result = df.groupby("A").cumsum(axis=1)
+    # GH 28549
+    # Good represention of when as_index=False is now behaving
+    # as expected
```
[the remark this answers is not in the document]
Comment (christopherzimmerman) [reply pandas/tests/groupby/test_function.py]: I was unsure what to do about this test.  There is a comment in the original test that says "cumsum is a transformer and should ignore as_index", but with the updated behavior, it respects `as_index` along the first axis.  I'm not sure what the actual behavior should be in this case.

**Comment by WillAyd** [comment]
Hmm there is still something weird going on here. On this branch I now see the following behavior:

```python
>>> df = pd.DataFrame({"a": [1, 1, 2, 2, 3, 3], "b": [1, 2, 3, 4, 5, 6]})
>>> df.groupby("a", as_index=False).shift()
     b
0  NaN
1  1.0
2  NaN
3  3.0
4  NaN
5  5.0
>>> df.groupby("a", as_index=False).apply(lambda x: x.shift())
     a    b
0  NaN  NaN
1  1.0  1.0
2  NaN  NaN
3  2.0  3.0
4  NaN  NaN
5  3.0  5.0
```

I would expect both of those outputs to be the same. Came across this looking at #13519 

We generally might need to step back and think about how the `as_index` parameter is presented to end users. Right now it seems to work for `agg` and `apply` but `transform` doesn't respect it. That might be OK, but at the same time I don't think we want people to start doing `apply(lambda x: x.shift())` instead of using `.shift()` if they do desire output with the grouper column(s) in tact

**Comment by jbrockmendel** [comment]
@christopherzimmerman can you rebase

**Comment by christopherzimmerman** [comment]
@WillAyd I tracked down the issue with `shift` that you mentioned.  Is that something that is worth a PR or should the behavior stay as it is now?

File: doc/source/whatsnew/v1.0.0.rst [hunk 12]
```diff
@@ -176,6 +176,48 @@ Backwards incompatible API changes
...
+
+- Previously, the result of :meth:`GroupBy.apply` sometimes contained the grouper column(s),
+  in both the index, and in the `DataFrame`. :meth:`GroupBy.apply`
+  now respects the ``as_index`` parameter, and only returns the grouper column(s) in
+  the result if ``as_index`` is set to `False`.  Other methods such as :meth:`GroupBy.resample`
+  exhibited similar behavior and now also respect the ``as_index`` parameter.
+
+*Previous Behavior*
+
+.. ipython:: python
```
Comment (WillAyd) [inline doc/source/whatsnew/v1.0.0.rst]: For previous behavior you need to convert this into a code-block and hard code the output, otherwise won't appear correctly

File: doc/source/whatsnew/v1.0.0.rst [hunk 13]
```diff
@@ -176,6 +176,48 @@ Backwards incompatible API changes
...
+
+    df = pd.DataFrame({"a": [1, 1, 2, 2, 3, 3], "b": [1, 2, 3, 4, 5, 6]})
+    df.groupby("a").apply(lambda x: x.sum())
+    df.groupby("a").apply(lambda x: x.iloc[0])
+    idx = pd.date_range('1/1/2000', periods=4, freq='T')
+    df = pd.DataFrame(data=4 * [range(2)],
+                      index=idx,
+                      columns=['a', 'b'])
+    df.iloc[2, 0] = 5
+    df.groupby('a').resample('M').sum()
```
Comment (WillAyd) [inline doc/source/whatsnew/v1.0.0.rst]: Can you use GroupBy.apply instead of resample? The link may not be obvious to all users

**Review by WillAyd — CHANGES_REQUESTED** [review]
OK thanks for identifying that. I think OK as a follow up PR

**Comment by jreback** [comment]
can you merge master

File: doc/source/whatsnew/v1.0.0.rst (hunk 13 above)
Comment (christopherzimmerman) [reply doc/source/whatsnew/v1.0.0.rst]: I added the resample bit due to one of Jeff's comments, since it's behavior is affected as well.  Should I turn it into it's own section and explain why the behavior changed?

**Review by WillAyd — APPROVED** [review]
lgtm @jreback

**Comment by jorisvandenbossche** [comment]
I am not fully convinced this is a "bug fix". 

The whole `as_index` / when is the group key set as index and when not or whether it is included in the group df or not, is a complete mess, for sure. 
But I think the possibly biggest change in this PR is in the dataframe that is passed to the applied function. While the whatsnew note mostly speaks about including the group key *in the result*, users can actually write functions right now that assume the group key is part of this dataframe on which the function operates.

Basically, that you can think of `df.groupby('a').apply(f)` as doing something like

```
for name, group in df.groupby('a'):
    res = f(group)
    ....
```

which is no longer true (as the `group` passed to `f` changed).

For reducing functions I agree that you typically don't want the group column to be included in the grouped dataframe on which to apply the function. But in general, for custom functions that are not necessarily reducing functions, this might not be necessarily true.

**Comment by jreback** [comment]
can you merge master and we'll take another look (and @jorisvandenbossche comments as well)

**Comment by WillAyd** [comment]
@christopherzimmerman looks like CI is red and some merge conflicts; any chance you can fix those up?

**Comment by christopherzimmerman** [comment]
@WillAyd I merged and fixed the failing tests.  It seems like tests keep getting written that rely on the old behavior of apply.

File: doc/source/whatsnew/v1.0.0.rst [hunk 14]
```diff
@@ -398,6 +398,81 @@ keywords.
 
    df.rename(index={0: 1}, columns={0: 2})
 
+
+.. _whatsnew_1000.api_breaking.GroupBy.apply:
```
Comment (jreback) [inline doc/source/whatsnew/v1.0.0.rst]: need to move to 1.1

File: doc/source/whatsnew/v1.0.0.rst [hunk 15]
```diff
@@ -398,6 +398,81 @@ keywords.
...
+  now respects the ``as_index`` parameter, and only returns the grouper column(s) in
+  the result if ``as_index`` is set to `False`.  Other methods such as :meth:`GroupBy.resample`
+  exhibited similar behavior and now also respect the ``as_index`` parameter.
+
+*Previous Behavior*
+
+.. code-block:: ipython
+
+    In [1]: df = pd.DataFrame({"a": [1, 1, 2, 2, 3, 3], "b": [1, 2, 3, 4, 5, 6]})
+
```
Comment (jreback) [inline doc/source/whatsnew/v1.0.0.rst]: show df here

File: doc/source/whatsnew/v1.0.0.rst [hunk 16]
```diff
@@ -398,6 +398,81 @@ keywords.
...
+    3  3  5
+
+    In [4]: idx = pd.date_range('1/1/2000', periods=4, freq='T')
+
+    In [5]: df = pd.DataFrame(data=4 * [range(2)],
+       ...:                   index=idx,
+       ...:                   columns=['a', 'b'])
+
+    In [6]: df.iloc[2, 0] = 5
+
```
Comment (jreback) [inline doc/source/whatsnew/v1.0.0.rst]: show df

File: doc/source/whatsnew/v1.0.0.rst [hunk 17]
```diff
@@ -398,6 +398,81 @@ keywords.
...
+    Out[7]:
+                  a  b
+    a
+    0 2000-01-31  0  3
+    5 2000-01-31  5  1
+
+
+*Current Behavior*
+
+.. ipython:: python
```
Comment (jreback) [inline doc/source/whatsnew/v1.0.0.rst]: break this up into 2 or more examples, its just too hard to follow like this.

meaning:

change1
before
new

change2
before
new

File: pandas/tests/groupby/test_categorical.py [hunk 18]
```diff
@@ -93,9 +93,16 @@ def f(x):
         return x.drop_duplicates("person_name").iloc[0]
 
     result = g.apply(f)
-    expected = x.iloc[[0, 1]].copy()
+
```
Comment (jreback) [inline pandas/tests/groupby/test_categorical.py]: so if the tests change a lot like this, make a new test

File: pandas/tests/groupby/test_categorical.py [hunk 19]
```diff
@@ -1246,6 +1253,16 @@ def test_get_nonexistent_category():
     # Accessing a Category that is not in the dataframe
     df = pd.DataFrame({"var": ["a", "a", "b", "b"], "val": range(4)})
     with pytest.raises(KeyError, match="'vau'"):
+        df.groupby("var").apply(
+            lambda rows: pd.DataFrame({"val": [rows.iloc[-1]["vau"]]})
+        )
+
+
+def test_category_as_grouper_keys(as_index):
+    # Accessing a key that is not in the dataframe
```
Comment (jreback) [inline pandas/tests/groupby/test_categorical.py]: wait this raises now?

File: pandas/tests/groupby/test_function.py [hunk 20]
```diff
@@ -103,19 +99,33 @@ def test_builtins_apply(keys, f):
...
         result,  # numpy's equivalent function
         df.groupby(keys).apply(getattr(np, fname)),
     )
 
     if f != sum:
-        expected = df.groupby(keys).agg(fname).reset_index()
-        expected.set_index(keys, inplace=True, drop=False)
+        # GH 28549
+        # No longer need to reset/set index here
+        expected = df.groupby(keys).agg(fname)
```
Comment (jreback) [inline pandas/tests/groupby/test_function.py]: can you try not to use groupby here and just construct the expected result.

the comment is also not very useful as a future reader won't have context.

can you break this out to a new test

File: pandas/tests/groupby/test_function.py [hunk 21]
```diff
@@ -340,10 +350,10 @@ def test_cython_api2():
     tm.assert_frame_equal(result, expected)
 
     # GH 13994
-    result = df.groupby("A").cumsum(axis=1)
+    result = df.groupby("A", as_index=False).cumsum(axis=1)
```
Comment (jreback) [inline pandas/tests/groupby/test_function.py]: can you test as_indexer=True as well. this seems like a big api change here

File: pandas/tests/groupby/test_groupby.py [hunk 22]
```diff
@@ -955,14 +956,15 @@ def f_no_copy(x):
...
 def test_no_mutate_but_looks_like():
-
     # GH 8467
     # first show's mutation indicator
     # second does not, but should yield the same results
     df = DataFrame({"key": [1, 1, 1, 2, 2, 2, 3, 3, 3], "value": range(9)})
 
-    result1 = df.groupby("key", group_keys=True).apply(lambda x: x[:].key)
-    result2 = df.groupby("key", group_keys=True).apply(lambda x: x.key)
+    result1 = df.groupby("key", group_keys=True, as_index=False).apply(
```
Comment (jreback) [inline pandas/tests/groupby/test_groupby.py]: what happens with as_index=True?

File: doc/source/whatsnew/v1.0.0.rst [hunk 23]
```diff
@@ -398,6 +398,81 @@ keywords.
...
+                      columns=['a', 'b'])
+    df.iloc[2, 0] = 5
+    df.groupby('a').resample('M').sum()
+
+
+All :class:`SeriesGroupBy` aggregation methods now respect the ``observed`` keyword
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+The following methods now also correctly output values for unobserved categories when called through ``groupby(..., observed=False)`` (:issue:`17605`)
+
+- :meth:`SeriesGroupBy.count`
```
Comment (jreback) [inline doc/source/whatsnew/v1.0.0.rst]: where are these tested? can you do this as a separate change?

**Review by jreback — CHANGES_REQUESTED** [review]
@christopherzimmerman there is a lot here. Is it possible to do a pre-cursor PR which adds the fixtures & locks down the current behavior, then this change PR will be more obvious what exactly is changing?

**Comment by christopherzimmerman** [comment]
@jreback I'll look into making a PR that is a bit less of an overhaul than this one.  Is the behavior that this PR introduces the desired behavior, with `apply` not containing the grouper columns unless specified?  If that *is* desired, then I'll also need to follow up with PRs to handle `shift`, `resample`, and several others so that they all respect the context as well.

I understand what @jorisvandenbossche is saying about this making some major changes, and I just want to make sure it's correct before spending too much more time on it.

**Comment by jreback** [comment]
@christopherzimmerman I think the behvaior changes in this PR *are* correct, meaning that these are some long standing bugs. However it would be better to do the non-controversial / easy ones first; the reason is we might need to deprecate some behavior rather than just breaking it. So cleanly separating these changes is key.

thanks for working on this!

File: pandas/tests/groupby/test_function.py [hunk 24]
```diff
@@ -88,10 +88,6 @@ def test_intercept_builtin_sum():
     tm.assert_series_equal(result2, expected)
 
 
-# @pytest.mark.parametrize("f", [max, min, sum])
-# def test_builtins_apply(f):
```
Comment (jbrockmendel) [inline pandas/tests/groupby/test_function.py]: this is the kind of thing that can be done in a separate PR

**Comment by jreback** [comment]
@christopherzimmerman happy to have a lesser scope PR to just fix the problem.

File: pandas/tests/groupby/test_categorical.py [hunk 25]
```diff
@@ -1246,6 +1253,16 @@ def test_get_nonexistent_category():
     # Accessing a Category that is not in the dataframe
     df = pd.DataFrame({"var": ["a", "a", "b", "b"], "val": range(4)})
     with pytest.raises(KeyError, match="'vau'"):
+        df.groupby("var").apply(
+            lambda rows: pd.DataFrame({"val": [rows.iloc[-1]["vau"]]})
```
Comment (jbrockmendel) [inline pandas/tests/groupby/test_categorical.py]: nitpick: if its the `.apply` that raises, let's do the df.groupby outside of the pytest.raises block, e.g. `gb = df.groupby("var")`

**Comment by jreback** [comment]
@christopherzimmerman if you can push a more limited scope i think we would like to fix these bugs

**Comment by jreback** [comment]
closing this as stale. as indicated above we would welcome this type of change, but we need to do it incrementally to avoid breaking the world.
